libsdk/libsdk.go

```diff
@@ -259,6 +260,142 @@ type client struct {
 	mu   sync.Mutex
 }
 
+//export cfs_get_dir_lock
+func cfs_get_dir_lock(id C.int64_t, path *C.char, lock_id *C.int64_t, valid_time **C.char) C.int {
+	c, exist := getClient(int64(id))
+	if !exist {
+		return statusEINVAL
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	log.LogInfof("cfs_get_dir_lock begin path(%s)\n", c.absPath(C.GoString(path)))
+
+	info, err := c.lookupPath(c.absPath(C.GoString(path)))
+	if err != nil {
+		return errorToStatus(err)
+	}
+
+	ino := info.Inode
+	name := "dir_lock"
+
+	dir_lock, err := c.getxattr(ino, name)
+	if err != nil {
+		log.LogErrorf("getxattr dir_lock failed, path(%s) ino(%v) err(%v)", c.absPath(C.GoString(path)), ino, err)
+		return errorToStatus(err)
+	}
+	if len(dir_lock) == 0 {
+		log.LogInfof("dir(%s) is not locked\n", c.absPath(C.GoString(path)))
+		return errorToStatus(syscall.ENOENT)
+	}
+
+	parts := strings.Split(string(dir_lock), "|")
+	lockIdStr := parts[0]
+	lease := parts[1]
+	lockId, _ := strconv.Atoi(lockIdStr)
+	*lock_id = C.int64_t(lockId)
+	*valid_time = C.CString(lease)
+
+	log.LogInfof("cfs_get_dir_lock end path(%s) lock_id(%d) lease(%s)\n", c.absPath(C.GoString(path)), lockId, lease)
+	return statusOK
+}
+
+//export cfs_lock_dir
+func cfs_lock_dir(id C.int64_t, path *C.char, interval C.int64_t, lock_id C.int64_t) C.int64_t {
+	c, exist := getClient(int64(id))
+	if !exist {
+		return C.int64_t(statusEINVAL)
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	log.LogInfof("cfs_lock_dir path(%s) interval(%d) lock_id(%d)\n", c.absPath(C.GoString(path)), interval, lock_id)
+
+	info, err := c.lookupPath(c.absPath(C.GoString(path)))
+	if err != nil {
+		return C.int64_t(errorToStatus(err))
+	}
+
+	ino := info.Inode
+	name := "dir_lock"
+	now := time.Now()
+
+	dir_lock, err := c.getxattr(ino, name)
+	if err != nil {
+		log.LogErrorf("getxattr dir_lock failed, path(%s) ino(%v) err(%v)", c.absPath(C.GoString(path)), ino, err)
+		return C.int64_t(errorToStatus(err))
+	}
+	log.LogInfof("getxattr dir_lock, path(%s) value(%s)\n", c.absPath(C.GoString(path)), dir_lock)
+
+	lockId := strconv.Itoa(int(lock_id))
+	renew_dir_lock := false
+	lock_expired := false
+	// check whether dir_lock has been setted by others
+	if len(dir_lock) != 0 {
+		parts := strings.Split(string(dir_lock), "|")
+		setted_lock_id := parts[0]
+		lease := parts[1]
+		lease_time, _ := time.Parse("2006-01-02 15:04:05", lease)
+		renew_dir_lock = (setted_lock_id == lockId)
+
+		// convert time before compare (CST/UTC)
+		now_time, _ := time.Parse("2006-01-02 15:04:05", now.Format("2006-01-02 15:04:05"))
```

Comment:
Storing time as a string using xattr consumes more space compared to storing it as an integer.